source/auth/auth.rst

```diff
@@ -1159,6 +1170,259 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.
+
+If the refresh callback is given and the request callback is not given,
+the driver MUST raise an error.  If DEVICE_NAME is given and one or more
+callbacks are given, the driver MUST raise an error.
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws",).
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drivers MUST support device workflows for "aws", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument and the second client step of the
+OIDC SASL exchange MUST be made directly, skipping the clientStep1.
+Drivers MUST raise an error if both a DEVICE_NAME and PRINCIPAL_NAME are
+given, since the device workflow will not use the PRINCIPAL_NAME.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
+interpret it as a file path.  The contents of the file are read as the
+access token.
+
+
+Caching Credentials
+```````````````````
+
+Drivers MUST enable caching when callback(s) are provided to the mongo client.
+When an authorization request is made and there is a valid cached response,
+the driver MUST use the cached authorization token if it has not expired.
+
+A cache value is considered valid if it has been created in the past 5 hours.
+The cache is kept alive to preserve the serverStep1 response, as well as
+account for refresh tokens, which typically have an (unknown) lifetime that
+is longer than the access token lifetime.  The refresh token may either be
+stored by the user application, or as part of the request/refresh token
+response in the driver cache. To prevent a memory leak, the driver MUST clear
+invalid cache values at a regular interval, or during every authentication
+attempt.
+
+The cache keys MUST include the principal name (or empty string) and the
+IP address and port for the current server's endpoint.  Using the IP address and port accounts for the case when two different servers use the same
+principal name but could be configurated differently.  There is an edge case
+where if the same principal name is used and two aliases to the same local
+host address are given, there will be duplicate user/device interactions,
+unless the driver can resolve the local host address as well.  Note that
+because we use the current host name, changing the primary host on a replica
+set will result in a cache miss.
```

Comment:
Does this still apply? If the cache is based on the server IP address, then it will have different entries for all RS members.

Reply:
I'll rephrase it, that is what I was trying to convey.